### PR TITLE
Add support for Nordic's nRF52810 IC

### PR DESCRIPTION
**Pull request by carlescufi**
Adds support for the IC itself and a new board which enforces the 52810 limitations but mirrors the nRF52 DK hardware, since there is no development kit specific to the 52810.

Tested with the RTT console since UART is not functional yet.

Note that BLE is not functional yet either. We require a new version of nrfx with the appropriate radio bindings and then @cvinayak will be able to add a layer for it in the bluetooth hal folder.

### CODE REVIEW

File: include/arch/arm/cortex_m/mpu/arm_mpu.h
Comment (carlescufi) [inline include/arch/arm/cortex_m/mpu/arm_mpu.h]: @galak @agross-linaro @ioannisg please review this change

**Comment by carlescufi** [comment]
CC @junqingzou @jarz-nordic @kl-cruz

File: arch/arm/soc/nordic_nrf/nrf52/Kconfig.defconfig.nrf52810_QFAA
Comment (ioannisg) [inline arch/arm/soc/nordic_nrf/nrf52/Kconfig.defconfig.nrf52810_QFAA]: Fix number

File: arch/arm/soc/nordic_nrf/nrf52/soc.c
Comment (jarz-nordic) [inline arch/arm/soc/nordic_nrf/nrf52/soc.c]: I think that both chips shall have different init functions. Most of PANs were fixed in nrf52810

File: boards/arm/nrf52810_pca10040/Kconfig
Comment (ioannisg) [inline boards/arm/nrf52810_pca10040/Kconfig]: pca10068 - also correct folder name

File: boards/arm/nrf52810_pca10040/Kconfig
Comment (ioannisg) [inline boards/arm/nrf52810_pca10040/Kconfig]: here too, and everywhere else

File: boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts
Comment (jarz-nordic) [inline boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts]: nrf52810 does not have an UART0, only UARTE0

File: dts/arm/nordic/nrf52810.dtsi
Comment (jarz-nordic) [inline dts/arm/nordic/nrf52810.dtsi]: nrf52810 does not have TWIM1.
It has only TWIM0 at address 40003000

**Review by ioannisg — CHANGES_REQUESTED** [review]

**Review by jakub-uC — CHANGES_REQUESTED** [review]

File: arch/arm/soc/nordic_nrf/nrf52/soc.c
Comment (ioannisg) [reply arch/arm/soc/nordic_nrf/nrf52/soc.c]: correct.

File: include/arch/arm/cortex_m/mpu/arm_mpu.h
Comment (ioannisg) [reply include/arch/arm/cortex_m/mpu/arm_mpu.h]: This is not correct. These macros are used to encode the region size in the MPU register, when configuring MPU regions, and AFAIK, we only have powers of two, there.
Check table B3-43 in ARMv7-M technical reference manual.

File: boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts
Comment (nvlsianpu) [inline boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts]: Why the scratch partition is of the image slot size? The size of 0x2000 will be better here (1/5 so, the same ratio as for nrf52832).

File: boards/arm/nrf52810_pca10040/Kconfig
Comment (carlescufi) [reply boards/arm/nrf52810_pca10040/Kconfig]: No, see commit messages for explanation

File: boards/arm/nrf52810_pca10040/Kconfig
Comment (carlescufi) [reply boards/arm/nrf52810_pca10040/Kconfig]: same

File: arch/arm/soc/nordic_nrf/nrf52/Kconfig.series
Comment (galak) [inline arch/arm/soc/nordic_nrf/nrf52/Kconfig.series]: At this point, seems like move the 'select CPU_HAS_FPU' over to the SOC_NRF52... would be better.

File: boards/arm/nrf52810_pca10040/board.h
Comment (galak) [inline boards/arm/nrf52810_pca10040/board.h]: Can we do this via DTS instead since this is a new board.

File: boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts
Comment (galak) [reply boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts]: While @carlescufi comment is correct, if its the case that UARTE0 can only be the 'E' form, than the compat in the SOC dts should just be 'nordic,nrf-uarte' and we don't need to update the compat in the board.dts

File: boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts
Comment (galak) [inline boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts]: See above comment if this compat should just be in the SoC dtsi

**Review by galak — DISMISSED** [review]
Looks good, a few dts and other style comments.

File: arch/arm/soc/nordic_nrf/nrf52/soc.c
Comment (carlescufi) [reply arch/arm/soc/nordic_nrf/nrf52/soc.c]: Fixed

File: boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts
Comment (carlescufi) [reply boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts]: It's already the case that compat in the SoC dts is only nrf-uarte

File: dts/arm/nordic/nrf52810.dtsi
Comment (carlescufi) [reply dts/arm/nordic/nrf52810.dtsi]: fixed

File: arch/arm/soc/nordic_nrf/nrf52/Kconfig.defconfig.nrf52810_QFAA
Comment (carlescufi) [reply arch/arm/soc/nordic_nrf/nrf52/Kconfig.defconfig.nrf52810_QFAA]: fixed

File: boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts
Comment (carlescufi) [reply boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts]: Fixed, thanks

File: arch/arm/soc/nordic_nrf/nrf52/Kconfig.series
Comment (carlescufi) [reply arch/arm/soc/nordic_nrf/nrf52/Kconfig.series]: Done

File: boards/arm/nrf52810_pca10040/board.h
Comment (carlescufi) [reply boards/arm/nrf52810_pca10040/board.h]: I would love to, but I think I would need some help here. Can you give me some pointers?

File: boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts
Comment (carlescufi) [reply boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts]: I removed this line from the board dts file, since it's already in the dtsi

File: boards/arm/nrf52810_pca10040/board.h
Comment (galak) [reply boards/arm/nrf52810_pca10040/board.h]: You'll also need to pull in some changes from PR #7444 (meaning adding the dt nodes for GPIO etc).

Comment (carlescufi) [reply boards/arm/nrf52810_pca10040/board.h]: Very well. In that case I will wait until #7444 is merged, since this should not be merged before 1.12 is released anyway. Thanks for the pointers.

**Review by ulfalizer — CHANGES_REQUESTED** [review]

File: ext/hal/nordic/nrfx_config_nrf52810.h
Comment (kl-cruz) [inline ext/hal/nordic/nrfx_config_nrf52810.h]: nRF52810 is not equipped with SPI. There are SPIM and SPIS

File: ext/hal/nordic/nrfx_config_nrf52810.h
Comment (kl-cruz) [inline ext/hal/nordic/nrfx_config_nrf52810.h]: There is no TIMER3 in nRF52810

File: ext/hal/nordic/nrfx_config_nrf52810.h
Comment (kl-cruz) [inline ext/hal/nordic/nrfx_config_nrf52810.h]: There is no TIMER4 in nRF52810

File: ext/hal/nordic/nrfx_config_nrf52810.h
Comment (kl-cruz) [inline ext/hal/nordic/nrfx_config_nrf52810.h]: no I2S in nRF52810

File: ext/hal/nordic/nrfx_config_nrf52810.h
Comment (kl-cruz) [inline ext/hal/nordic/nrfx_config_nrf52810.h]: no LPCOMP in nRF52810

File: ext/hal/nordic/nrfx_config_nrf52810.h
Comment (kl-cruz) [inline ext/hal/nordic/nrfx_config_nrf52810.h]: only PWM0 in nRF52810

File: ext/hal/nordic/nrfx_config_nrf52810.h
Comment (kl-cruz) [inline ext/hal/nordic/nrfx_config_nrf52810.h]: only SPIM0 in 52810

File: ext/hal/nordic/nrfx_config_nrf52810.h
Comment (kl-cruz) [inline ext/hal/nordic/nrfx_config_nrf52810.h]: only TWIM0 in nRF52810

File: ext/hal/nordic/nrfx_config_nrf52810.h
Comment (kl-cruz) [inline ext/hal/nordic/nrfx_config_nrf52810.h]: no TWI in nRF52810.
This config file needs to be changed

**Review by kl-cruz — CHANGES_REQUESTED** [review]
config file needs rework

File: arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc
Comment (pizi-nordic) [inline arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc]: There is one more difference between nRF52810 and other members of the nRF52 family: The nRF52810 does not have instruction cache, which should be enabled on other SoCs.

At the moment it looks that we are not enabling cache at all. IMHO we should fix that asap as without cache performance is significantly degraded. Please fix that in this PR or create new issue.

**Review by pizi-nordic — APPROVED** [review]

File: arch/arm/soc/nordic_nrf/nrf52/Kconfig.defconfig.nrf52810_QFAA
Comment (carlescufi) [reply arch/arm/soc/nordic_nrf/nrf52/Kconfig.defconfig.nrf52810_QFAA]: Fixed, thanks

File: ext/hal/nordic/nrfx_config_nrf52810.h
Comment (carlescufi) [reply ext/hal/nordic/nrfx_config_nrf52810.h]: I pushed a new file, please re-review

File: ext/hal/nordic/nrfx_config_nrf52810.h
Comment (carlescufi) [reply ext/hal/nordic/nrfx_config_nrf52810.h]: I pushed a new file, please re-review

File: boards/arm/nrf52810_pca10040/Kconfig.defconfig
Comment (ulfalizer) [inline boards/arm/nrf52810_pca10040/Kconfig.defconfig]: Still missing quotes here.

**Review by ulfalizer — CHANGES_REQUESTED** [review]

File: boards/arm/nrf52810_pca10040/Kconfig.defconfig
Comment (carlescufi) [reply boards/arm/nrf52810_pca10040/Kconfig.defconfig]: Fixed

**Review by ulfalizer — APPROVED** [review]
Quoting nit addressed

File: arch/arm/soc/nordic_nrf/nrf52/mpu_regions.c
Comment (ioannisg) [inline arch/arm/soc/nordic_nrf/nrf52/mpu_regions.c]: use #if defined(.....)

**Review by ioannisg — CHANGES_REQUESTED** [review]

**Comment by ioannisg** [comment]
@carlescufi MPU regions' configuration looks good, now. 
I left some styling comments only.

File: dts/arm/nordic/nrf52810.dtsi
Comment (anangl) [inline dts/arm/nordic/nrf52810.dtsi]: nRF52810 has only TWIM and no TWI that is present on nRF51 and other nRF52s. Shouldn't a new binding ("nordic,nrf-twim") be introduced and this value changed?

**Review by dbkinder — CHANGES_REQUESTED** [review]
small tweak to the doc

File: ext/hal/nordic/nrfx_config_nrf52810.h
Comment (kl-cruz) [inline ext/hal/nordic/nrfx_config_nrf52810.h]: When this PR is merged (https://github.com/zephyrproject-rtos/zephyr/pull/7788) it should be changed to CONFIG_NRFX_TWIM like i the SPIM section. Line 1665 too.

File: tests/kernel/sched/schedule_api/src/test_priority_scheduling.c
Comment (ioannisg) [inline tests/kernel/sched/schedule_api/src/test_priority_scheduling.c]: Is this intended, @carlescufi ?

File: arch/arm/soc/nordic_nrf/nrf52/soc.c
Comment (ioannisg) [inline arch/arm/soc/nordic_nrf/nrf52/soc.c]: @anangl take a look here: I simply added nRF52810 nrf header, in soc.c.
Is there anything else, (besides defining DEVELOP_IN_NRF52832) that we need to do here?

Comment (anangl) [reply arch/arm/soc/nordic_nrf/nrf52/soc.c]: Good. But the line above should be `#if defined(CONFIG_SOC_NRF52810)`.
And that should be it.

Comment (ioannisg) [reply arch/arm/soc/nordic_nrf/nrf52/soc.c]: ups, my bad, thanks for catching this.
Will fix.

File: arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc
Comment (anangl) [inline arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc]: This addition is no longer needed (see 2 lines below).

File: arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc
Comment (anangl) [inline arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc]: This addition is no longer needed (see 2 lines below).

File: boards/arm/nrf52810_pca10040/CMakeLists.txt
Comment (anangl) [inline boards/arm/nrf52810_pca10040/CMakeLists.txt]: We used to write nrfx as nrfx. I think we should stay with this convention.
Anyway, I think it would be better to refer directly to the "system_nrf52810.c" file, instead of pointing to nrfx or MDK that resides inside it.

**Review by anangl — CHANGES_REQUESTED** [review]

File: arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc
Comment (ioannisg) [reply arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc]: thanks.

File: arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc
Comment (ioannisg) [reply arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc]: thanks

File: arch/arm/soc/nordic_nrf/nrf52/mpu_regions.c
Comment (ioannisg) [reply arch/arm/soc/nordic_nrf/nrf52/mpu_regions.c]: Fixed.

File: boards/arm/nrf52810_pca10040/Kconfig
Comment (ioannisg) [reply boards/arm/nrf52810_pca10040/Kconfig]: ok.

File: boards/arm/nrf52810_pca10040/Kconfig
Comment (ioannisg) [reply boards/arm/nrf52810_pca10040/Kconfig]: By the way @carlescufi do we really need these empty Kconfig files?

File: boards/arm/nrf52810_pca10040/Kconfig.defconfig
Comment (ioannisg) [inline boards/arm/nrf52810_pca10040/Kconfig.defconfig]: Is this needed @anangl ? Is 810 going to have only UARTE, instead?

File: boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts
Comment (ioannisg) [reply boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts]: then i suppose we are ok.

File: boards/arm/nrf52810_pca10040/doc/nrf52810_pca10040.rst
Comment (ioannisg) [reply boards/arm/nrf52810_pca10040/doc/nrf52810_pca10040.rst]: fixed

File: boards/arm/nrf52810_pca10040/Kconfig.defconfig
Comment (anangl) [reply boards/arm/nrf52810_pca10040/Kconfig.defconfig]: It is needed. It enables nrfx drivers for UARTs in general. The type of the driver (UART or UARTE) is selected with a different option.

Comment (ioannisg) [reply boards/arm/nrf52810_pca10040/Kconfig.defconfig]: ok

File: boards/arm/nrf52810_pca10040/doc/nrf52810_pca10040.rst
Comment (jarz-nordic) [inline boards/arm/nrf52810_pca10040/doc/nrf52810_pca10040.rst]: I am wondering if we shall mention here that this solution combines limations of nrf52832 and nrf52810.
User shall be aware that PANs fixed in 52810 are still valid with this approach, hence intialization time will be extended.
Furthermore nr52810 has 10bit transfer buffers for SPIM, TWIM, UARTE but you can use only 8 bits because it is HW limation in nr52832.

Comment (ioannisg) [reply boards/arm/nrf52810_pca10040/doc/nrf52810_pca10040.rst]: I think this is a rather generic comment, and we should not care about the details. I think we just need to stress that 810 is a reduced version of 832.

**Review by ioannisg — APPROVED** [review]

File: ext/hal/nordic/nrfx_config_nrf52810.h
Comment (ioannisg) [reply ext/hal/nordic/nrfx_config_nrf52810.h]: Will be addressed, thanks.

File: arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc
Comment (anangl) [inline arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc]: The same should be done for `config NFCT_PINS_AS_GPIOS` - this option does not make sense for nRF52810.

Comment (ioannisg) [reply arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc]: Will do.

**Review by anangl — APPROVED** [review]

**Review by kl-cruz — APPROVED** [review]

**Review by jakub-uC — APPROVED** [review]

**Review by nvlsianpu — APPROVED** [review]

**Comment by anangl** [comment]
Since #7334 is now in master, "nrfx_config_nrf52810.h" needs one more update: [this](https://github.com/zephyrproject-rtos/zephyr/pull/7334/files#diff-a56c8d031d7dadfe354891d51ab4b4f5L959) and [this](https://github.com/zephyrproject-rtos/zephyr/pull/7334/files#diff-a56c8d031d7dadfe354891d51ab4b4f5L965) change.
Hence, maybe it would make sense to merge #8525 first and then update commit f85cfced1a8b39bd5dd2335a98f322075131be4c regarding "nrfx_config_nrf52810.h" and "CMakeLists.txt"?

**Comment by carlescufi** [comment]
@anangl addressed, thank you

**Comment by carlescufi** [comment]
recheck

**Comment by carlescufi** [comment]
recheck

**Comment by carlescufi** [comment]
@dbkinder can you revisit this one?

**Comment by agross-linaro** [comment]
looks good to me

**Review by agross-oss — APPROVED** [review]

**Comment by carlescufi** [comment]
@galak @dbkinder will you be able to take a look at this soon? Your requests have been addressed so unless I hear back soon I will need to dismiss the reviews in order to get this merged. Thanks!

**Review by dbkinder — APPROVED** [review]

**Comment by carlescufi** [comment]
thanks @dbkinder!

**Pull request merged**
